File: typescript/src/lifecycle.ts
/**
 * License lifecycle state machine.
 *
 * The license goes through these states:
 *
 *                 ┌───────────┐
 *                 │  pending  │
 *                 └─────┬─────┘
 *                       │ activate()
 *                       ▼
 *      ┌──────────┐   ┌────────┐   ┌─────────┐   ┌─────────┐
 *      │ suspend  │◀──│ active │──▶│  grace  │──▶│ expired │
 *      │          │──▶│        │   │ (auto)  │   │ (auto)  │
 *      └──────────┘ r │        │   └─────────┘   └─────────┘
 *            │      e │        │         (past grace_until,
 *   suspend  │      s │        │          tick() fires)
 *    (any    │      u │        │
 *     non-   │      m │        │
 *     rev.)  │      e │        │
 *            ▼        │        │
 *      ┌──────────┐   │        │
 *      │ revoked  │◀──┴────────┘ revoke()  (terminal)
 *      └──────────┘
 *
 * Rules:
 *   - `pending → active` via `activate()` or first successful `registerUsage`.
 *   - `active ↔ suspended` via `suspend` / `resume`.
 *   - `active → grace` is computed, not persisted: reading the license when
 *     now ∈ [expires_at, grace_until) reports `grace`. Persisting grace via
 *     `tick()` on a scheduled sweep is also supported.
 *   - `grace → expired` persists via `expire()` after grace_until passes.
 *   - `* → revoked` via `revoke()`; terminal (no further transitions allowed).
 *   - `expired → active` via `renew()` (sets new expires_at/grace_until).
 *   - Every successful transition writes a single audit row in the same tx.
 *
 * The functions below take a `StorageTx` handle and MUST be called inside
 * `storage.withTransaction(...)`. They are transaction-atomic: the license
 * row is updated and the audit row appended in the same atomic unit.
 */

import { errors } from './errors.ts';
import type { Clock } from './id.ts';
import type { StorageTx } from './storage/types.ts';
import type { JSONValue, License, LicenseStatus, UUIDv7 } from './types.ts';

/** Actor attribution for audit log. Defaults to `'system'` when omitted. */
export interface TransitionOptions {
  readonly actor?: string;
}

/** Options for `renew`. Caller supplies the new end timestamps explicitly —
 *  lifecycle has no opinion on trial/subscription duration math. */
export interface RenewOptions extends TransitionOptions {
  readonly expires_at: string | null;
  readonly grace_until?: string | null;
}

// ---------- effective-status helpers ----------

/**
 * Compute the *effective* status of a license given a clock instant, without
 * mutating storage. Used by read paths that want to display/return "grace"
 * when the license has passed `expires_at` but is still within `grace_until`,
 * even if the persisted `status` column hasn't been flipped by a sweeper.
 */
export function effectiveStatus(license: License, nowIso: string): LicenseStatus {
  // Terminal / non-time-dependent states pass through untouched.
  if (license.status === 'revoked') return 'revoked';
  if (license.status === 'suspended') return 'suspended';
  if (license.status === 'pending') return 'pending';
  if (license.status === 'expired') return 'expired';

  // Active or grace — check expiry boundaries.
  const { expires_at, grace_until } = license;
  if (expires_at !== null && nowIso >= expires_at) {
    if (grace_until !== null && nowIso < grace_until) return 'grace';
    return 'expired';
  }
  return license.status; // 'active' or already-'grace'
}

// ---------- transitions ----------

/** `pending → active`. No-op (returns license unchanged) if already active. */
export async function activate(
  tx: StorageTx,
  license: License,
  clock: Clock,
  opts: TransitionOptions = {},
): Promise<License> {
  assertNotRevoked(license);
  if (license.status === 'active') return license;
  if (license.status !== 'pending') {
    throw illegalTransition(license.status, 'active');
  }
  const now = clock.nowIso();
  const updated = await tx.updateLicense(license.id, {
    status: 'active',
    activated_at: now,
  });
  await writeAudit(tx, license, updated, 'license.activated', now, opts.actor);
  return updated;
}

/** `active → suspended`. Revoked is rejected; any other non-revoked state is
 *  accepted (suspending an already-suspended license is a no-op). */
export async function suspend(
  tx: StorageTx,
  license: License,
  clock: Clock,
  opts: TransitionOptions = {},
): Promise<License> {
  assertNotRevoked(license);
  if (license.status === 'suspended') return license;
  const now = clock.nowIso();
  const updated = await tx.updateLicense(license.id, { status: 'suspended' });
  await writeAudit(tx, license, updated, 'license.suspended', now, opts.actor);
  return updated;
}

/** `suspended → active`. */
export async function resume(
  tx: StorageTx,
  license: License,
  clock: Clock,
  opts: TransitionOptions = {},
): Promise<License> {
  assertNotRevoked(license);
  if (license.status !== 'suspended') {
    throw illegalTransition(license.status, 'active');
  }
  const now = clock.nowIso();
  const updated = await tx.updateLicense(license.id, { status: 'active' });
  await writeAudit(tx, license, updated, 'license.resumed', now, opts.actor);
  return updated;
}

/** Terminal `* → revoked`. Revoking an already-revoked license is a no-op. */
export async function revoke(
  tx: StorageTx,
  license: License,
  clock: Clock,
  opts: TransitionOptions = {},
): Promise<License> {
  if (license.status === 'revoked') return license;
  const now = clock.nowIso();
  const updated = await tx.updateLicense(license.id, { status: 'revoked' });
  await writeAudit(tx, license, updated, 'license.revoked', now, opts.actor);
  return updated;
}

/** `active|grace → expired`. Used by scheduled sweepers after `grace_until`. */
export async function expire(
  tx: StorageTx,
  license: License,
  clock: Clock,
  opts: TransitionOptions = {},
): Promise<License> {
  assertNotRevoked(license);
  if (license.status === 'expired') return license;
  if (license.status === 'suspended' || license.status === 'pending') {
    throw illegalTransition(license.status, 'expired');
  }
  const now = clock.nowIso();
  const updated = await tx.updateLicense(license.id, { status: 'expired' });
  await writeAudit(tx, license, updated, 'license.expired', now, opts.actor);
  return updated;
}

/** `expired|active|grace → active` with new end timestamps. Suspended and
 *  revoked licenses cannot be renewed — resume or unrevoke-is-not-allowed. */
export async function renew(
  tx: StorageTx,
  license: License,
  clock: Clock,
  opts: RenewOptions,
): Promise<License> {
  assertNotRevoked(license);
  if (license.status === 'suspended') {
    throw illegalTransition(license.status, 'active');
  }
  const now = clock.nowIso();
  const patch: Parameters<StorageTx['updateLicense']>[1] = {
    status: 'active',
    expires_at: opts.expires_at,
  };
  // Only include `grace_until` when the caller passed it — otherwise we leave
  // the existing value in place.
  const updated = await tx.updateLicense(
    license.id,
    opts.grace_until !== undefined ? { ...patch, grace_until: opts.grace_until } : patch,
  );
  await writeAudit(tx, license, updated, 'license.renewed', now, opts.actor);
  return updated;
}

/**
 * "Tick" the license — persist any time-driven transition that the effective
 * status implies. Call this on a scheduled sweep. Returns the (possibly
 * updated) license.
 *
 * - `active` whose `expires_at` is past and `grace_until` is in the future →
 *   persist `grace`.
 * - `active`/`grace` whose `grace_until` is past → persist `expired`.
 */
export async function tick(
  tx: StorageTx,
  license: License,
  clock: Clock,
  opts: TransitionOptions = {},
): Promise<License> {
  const now = clock.nowIso();
  const target = effectiveStatus(license, now);
  if (target === license.status) return license;
  if (target === 'grace') {
    const updated = await tx.updateLicense(license.id, { status: 'grace' });
    await writeAudit(tx, license, updated, 'license.grace_entered', now, opts.actor);
    return updated;
  }
  if (target === 'expired') {
    const updated = await tx.updateLicense(license.id, { status: 'expired' });
    await writeAudit(tx, license, updated, 'license.expired', now, opts.actor);
    return updated;
  }
  return license;
}

// ---------- internals ----------

function assertNotRevoked(license: License): void {
  if (license.status === 'revoked') throw errors.licenseRevoked();
}

function illegalTransition(from: LicenseStatus, to: LicenseStatus): Error {
  return errors.illegalLifecycleTransition(from, to);
}

async function writeAudit(
  tx: StorageTx,
  prior: License,
  next: License,
  event: string,
  occurredAt: string,
  actor: string | undefined,
): Promise<void> {
  await tx.appendAudit({
    license_id: prior.id,
    scope_id: prior.scope_id,
    actor: actor ?? 'system',
    event,
    prior_state: stateSnapshot(prior),
    new_state: stateSnapshot(next),
    occurred_at: occurredAt,
  });
}

function stateSnapshot(l: License): Readonly<Record<string, JSONValue>> {
  return {
    status: l.status,
    activated_at: l.activated_at,
    expires_at: l.expires_at,
    grace_until: l.grace_until,
  };
}

// Export the type with a self-reference helper the tests may want.
export type { UUIDv7 };

File: licensing/lifecycle.go
package licensing

// License lifecycle state machine — Go port of
// typescript/packages/core/src/lifecycle.ts.
//
// State diagram:
//
//	pending → active ↔ suspended
//	active → grace → expired
//	* → revoked (terminal)
//
// "grace" is computed, not persisted: effectiveStatus() dynamically
// determines if a license is in grace (between expires_at and grace_until)
// without a database write. The tick() function can optionally persist this.
//
// Every transition writes a single audit row in the same transaction.
// Functions take a StorageTx handle and MUST be called inside
// Storage.WithTransaction.

// TransitionOptions carries optional actor attribution for audit log entries.
type TransitionOptions struct {
	// Actor defaults to "system" when empty.
	Actor string
}

// RenewOptions extends TransitionOptions with new end timestamps.
type RenewOptions struct {
	ExpiresAt  *string
	GraceUntil *string
	TransitionOptions
	GraceUntilSet bool
}

// ---------- effective-status helper ----------

// EffectiveStatus computes the effective status of a license given a
// clock instant, without mutating storage. Used by read paths that want
// to display "grace" when the license has passed expires_at but is still
// within grace_until, even if the persisted status hasn't been flipped.
func EffectiveStatus(license *License, nowISO string) LicenseStatus {
	switch license.Status {
	case LicenseStatusRevoked, LicenseStatusSuspended,
		LicenseStatusPending, LicenseStatusExpired:
		return license.Status
	}

	// Active or grace — check expiry boundaries.
	if license.ExpiresAt != nil && nowISO >= *license.ExpiresAt {
		if license.GraceUntil != nil && nowISO < *license.GraceUntil {
			return LicenseStatusGrace
		}
		return LicenseStatusExpired
	}
	return license.Status // "active" or already "grace"
}

// ---------- transitions ----------

// Activate transitions pending → active. No-op if already active.
func Activate(tx StorageTx, license *License, clock Clock, opts TransitionOptions) (*License, error) {
	if err := assertNotRevoked(license); err != nil {
		return nil, err
	}
	if license.Status == LicenseStatusActive {
		return license, nil
	}
	if license.Status != LicenseStatusPending {
		return nil, illegalTransition(license.Status, LicenseStatusActive)
	}
	now := clock.NowISO()
	updated, err := tx.UpdateLicense(license.ID, LicensePatch{
		Status:      ptr(LicenseStatusActive),
		ActivatedAt: OptString{Set: true, Value: &now},
	})
	if err != nil {
		return nil, err
	}
	if err := writeLifecycleAudit(tx, license, updated, "license.activated", now, opts.Actor); err != nil {
		return nil, err
	}
	return updated, nil
}

// Suspend transitions active → suspended. Revoked is rejected; suspending
// an already-suspended license is a no-op.
func Suspend(tx StorageTx, license *License, clock Clock, opts TransitionOptions) (*License, error) {
	if err := assertNotRevoked(license); err != nil {
		return nil, err
	}
	if license.Status == LicenseStatusSuspended {
		return license, nil
	}
	now := clock.NowISO()
	updated, err := tx.UpdateLicense(license.ID, LicensePatch{
		Status: ptr(LicenseStatusSuspended),
	})
	if err != nil {
		return nil, err
	}
	if err := writeLifecycleAudit(tx, license, updated, "license.suspended", now, opts.Actor); err != nil {
		return nil, err
	}
	return updated, nil
}

// Resume transitions suspended → active.
func Resume(tx StorageTx, license *License, clock Clock, opts TransitionOptions) (*License, error) {
	if err := assertNotRevoked(license); err != nil {
		return nil, err
	}
	if license.Status != LicenseStatusSuspended {
		return nil, illegalTransition(license.Status, LicenseStatusActive)
	}
	now := clock.NowISO()
	updated, err := tx.UpdateLicense(license.ID, LicensePatch{
		Status: ptr(LicenseStatusActive),
	})
	if err != nil {
		return nil, err
	}
	if err := writeLifecycleAudit(tx, license, updated, "license.resumed", now, opts.Actor); err != nil {
		return nil, err
	}
	return updated, nil
}

// Revoke transitions * → revoked. Terminal. Revoking an already-revoked
// license is a no-op.
func Revoke(tx StorageTx, license *License, clock Clock, opts TransitionOptions) (*License, error) {
	if license.Status == LicenseStatusRevoked {
		return license, nil
	}
	now := clock.NowISO()
	updated, err := tx.UpdateLicense(license.ID, LicensePatch{
		Status: ptr(LicenseStatusRevoked),
	})
	if err != nil {
		return nil, err
	}
	if err := writeLifecycleAudit(tx, license, updated, "license.revoked", now, opts.Actor); err != nil {
		return nil, err
	}
	return updated, nil
}

// Expire transitions active|grace → expired. Used by scheduled sweepers
// after grace_until.
func Expire(tx StorageTx, license *License, clock Clock, opts TransitionOptions) (*License, error) {
	if err := assertNotRevoked(license); err != nil {
		return nil, err
	}
	if license.Status == LicenseStatusExpired {
		return license, nil
	}
	if license.Status == LicenseStatusSuspended || license.Status == LicenseStatusPending {
		return nil, illegalTransition(license.Status, LicenseStatusExpired)
	}
	now := clock.NowISO()
	updated, err := tx.UpdateLicense(license.ID, LicensePatch{
		Status: ptr(LicenseStatusExpired),
	})
	if err != nil {
		return nil, err
	}
	if err := writeLifecycleAudit(tx, license, updated, "license.expired", now, opts.Actor); err != nil {
		return nil, err
	}
	return updated, nil
}

// Renew transitions expired|active|grace → active with new end timestamps.
// Suspended and revoked licenses cannot be renewed.
func Renew(tx StorageTx, license *License, clock Clock, opts RenewOptions) (*License, error) {
	if err := assertNotRevoked(license); err != nil {
		return nil, err
	}
	if license.Status == LicenseStatusSuspended {
		return nil, illegalTransition(license.Status, LicenseStatusActive)
	}
	now := clock.NowISO()
	patch := LicensePatch{
		Status:    ptr(LicenseStatusActive),
		ExpiresAt: OptString{Set: true, Value: opts.ExpiresAt},
	}
	if opts.GraceUntilSet {
		patch.GraceUntil = OptString{Set: true, Value: opts.GraceUntil}
	}
	updated, err := tx.UpdateLicense(license.ID, patch)
	if err != nil {
		return nil, err
	}
	if err := writeLifecycleAudit(tx, license, updated, "license.renewed", now, opts.Actor); err != nil {
		return nil, err
	}
	return updated, nil
}

// Tick persists any time-driven transition that EffectiveStatus implies.
// Call on a scheduled sweep.
//
//   - active whose expires_at is past and grace_until is future → persist grace
//   - active|grace whose grace_until is past → persist expired
func Tick(tx StorageTx, license *License, clock Clock, opts TransitionOptions) (*License, error) {
	now := clock.NowISO()
	target := EffectiveStatus(license, now)
	if target == license.Status {
		return license, nil
	}
	switch target {
	case LicenseStatusGrace:
		updated, err := tx.UpdateLicense(license.ID, LicensePatch{
			Status: ptr(LicenseStatusGrace),
		})
		if err != nil {
			return nil, err
		}
		if err := writeLifecycleAudit(tx, license, updated, "license.grace_entered", now, opts.Actor); err != nil {
			return nil, err
		}
		return updated, nil

	case LicenseStatusExpired:
		updated, err := tx.UpdateLicense(license.ID, LicensePatch{
			Status: ptr(LicenseStatusExpired),
		})
		if err != nil {
			return nil, err
		}
		if err := writeLifecycleAudit(tx, license, updated, "license.expired", now, opts.Actor); err != nil {
			return nil, err
		}
		return updated, nil
	}
	return license, nil
}

// ---------- internals ----------

func assertNotRevoked(license *License) error {
	if license.Status == LicenseStatusRevoked {
		return newError(CodeLicenseRevoked, "license is revoked", nil)
	}
	return nil
}

func illegalTransition(from, to LicenseStatus) error {
	return newError(CodeIllegalLifecycleTransition,
		"illegal lifecycle transition: "+string(from)+" → "+string(to),
		map[string]any{"from": string(from), "to": string(to)})
}

func writeLifecycleAudit(tx StorageTx, prior, next *License, event, occurredAt, actor string) error {
	if actor == "" {
		actor = "system"
	}
	_, err := tx.AppendAudit(AuditLogInput{
		LicenseID:  &prior.ID,
		ScopeID:    prior.ScopeID,
		Actor:      actor,
		Event:      event,
		PriorState: lifecycleSnapshot(prior),
		NewState:   lifecycleSnapshot(next),
		OccurredAt: occurredAt,
	})
	return err
}

func lifecycleSnapshot(l *License) map[string]any {
	m := map[string]any{
		"status": string(l.Status),
	}
	if l.ActivatedAt != nil {
		m["activated_at"] = *l.ActivatedAt
	} else {
		m["activated_at"] = nil
	}
	if l.ExpiresAt != nil {
		m["expires_at"] = *l.ExpiresAt
	} else {
		m["expires_at"] = nil
	}
	if l.GraceUntil != nil {
		m["grace_until"] = *l.GraceUntil
	} else {
		m["grace_until"] = nil
	}
	return m
}

func ptr[T any](v T) *T { return &v }
